activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,153 @@
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects.
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`,
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      DEFAULT_KEY = Object.new
+      CACHE_METHOD_OBJ = ActiveSupport::Cache::Store.new
+      @@has_rails_cache = :unknown
+
+      def initialize(key = DEFAULT_KEY)
+        @cache_key = nil
+        @cache_version = nil
+        @key_parts = []
+        @cache_method_obj = cache_method_obj
+
+        self << key unless key == DEFAULT_KEY
+      end
+
+      def cache_key
+        @cache_key ||= @cache_method_obj.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_method_obj.send(:expanded_version, @key_parts)
+
+        if @cache_version == "/"
+          nil
+        else
+          @cache_version
+        end
+      end
+
+      def cache_key_with_version
+        @cache_method_obj.send(:retrieve_cache_key, @key_parts)
+      end
+
+      def update(key)
+        @key_parts << key
+        @cache_key = nil
+        @cache_version = nil
+        self
+      end
+      alias :<< :update
+
+      def length
+        cache_key.length
+      end
+
+      private
+        def cache_method_obj
+          if @@has_rails_cache == :unknown
+            @@has_rails_cache = defined?(Rails) && Rails.respond_to?(:cache)
+          end
+
+          if @@has_rails_cache
+            Rails.cache
+          else
+            CACHE_METHOD_OBJ
+          end
+        end
```

Comment:
I like using dependency injection, switched to an optional second argument that defaults to grabbing the information from the class.

I tried to inject it in `collection_caching.rb` directly but when I do I get several failures since Rails isn't guaranteed to exist and have a `cache` object on it in the actionview library

```
Error:
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching:
NoMethodError: undefined method `cache' for Rails:Module
```

Reply:
You shouldn't call `Rails.cache` there, we've already set up a `collection_cache` to use there which fetches what's basically `Rails.cache` via the Railtie: https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/railtie.rb#L96

We should convert this to `mattr_accessor` ala https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb#L10 and remove the default_cache_object method.